Description: address a typo in the source package
Author: tony mancill <tmancill@debian.org>
Forwarded: no

--- a/src/kj/filesystem-disk-unix.c++
+++ b/src/kj/filesystem-disk-unix.c++
@@ -1094,7 +1094,7 @@
           // Presumably because the target path doesn't exist.
           if (has(mode, WriteMode::CREATE)) {
             KJ_FAIL_ASSERT("rename(tmp, path) claimed path exists but "
-                "renameat2(fromPath, toPath, EXCAHNGE) said it doest; concurrent modification?",
+                "renameat2(fromPath, toPath, EXCHANGE) said it doesn't; concurrent modification?",
                 fromPath, toPath) { return false; }
           } else {
             // Assume target doesn't exist.
